server/src/com/cloud/storage/snapshot/SnapshotManagerImpl.java

```diff
@@ -1123,6 +1150,39 @@ public SnapshotInfo takeSnapshot(VolumeInfo volume) throws ResourceAllocationExc
         return snapshot;
     }
 
+    protected class BackupSnapshotTask extends ManagedContextRunnable {
+        SnapshotInfo snapshot;
+        int maxAttempts;
+        SnapshotStrategy snapshotStrategy;
+
+        public BackupSnapshotTask(SnapshotInfo snap, int maxRetries, SnapshotStrategy strategy) {
+            snapshot = snap;
+            maxAttempts = maxRetries;
+            snapshotStrategy = strategy;
+        }
+
+        @Override
+        protected void runInContext() {
+            try {
+                s_logger.debug("Value of attempts is " + maxAttempts);
+
+                SnapshotInfo backupedSnapshot = snapshotStrategy.backupSnapshot(snapshot);
+
+                if (backupedSnapshot != null) {
+                    snapshotStrategy.cleanSnapshotAfterCreation(snapshot);
+                }
+            } catch (final Exception e) {
+                if (maxAttempts >= 0) {
+                    s_logger.debug("Attempt failed, in retrying backup");
+                    backupSnapshotExecutor.schedule(new BackupSnapshotTask(snapshot, --maxAttempts, snapshotStrategy), retryInterval, TimeUnit.SECONDS);
+                } else {
+                    s_logger.debug("Done with attempts, with backing up");
```

Comment:
Again add more information about the snapshot id, how many retries were attempted etc.